pkg/catalog/walk_entry_iterator.go

```diff
@@ -0,0 +1,156 @@
+package catalog
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/treeverse/lakefs/pkg/ingest/store"
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type walkEntryIterator struct {
+	entries chan EntryWithMarker
+	walker  *store.WalkerWrapper
+
+	done   chan bool
+	closed *atomic.Bool
+	err    *atomic.Error
+
+	curr EntryWithMarker
+}
+
+// Mark stands for pagination information when listing objects from the blockstore.
+// It is used for server-client communication on the status of range ingestion.
+type Mark store.Mark
+
+type EntryWithMarker struct {
+	EntryRecord
+	Mark
+}
+
+var (
+	// ErrItClosed is used to determine the reason for the end of the walk
+	ErrItClosed = errors.New("iterator closed")
+
+	errSeekGENotSupported = errors.New("SeekGE not supported for walk entry iterator")
+)
+
+// buffer size of the buffer between reading entries from the blockstore Walk and passing it on
+const bufferSize = 100
+
+// WalkerFactory provides an abstraction for creating Walker
+type WalkerFactory interface {
+	GetWalker(ctx context.Context, opts store.WalkerOptions) (*store.WalkerWrapper, error)
+}
+
+func NewWalkEntryIterator(ctx context.Context, factory WalkerFactory, fromSourceURI, prepend, after, continuationToken string) (*walkEntryIterator, error) {
```

Comment:
Question - if the iterator creates just one walker, why does it need a factory? We can just pass it a walker

Reply:
Right, done.